# Python Date and Math Modules

## Practice 

In [22]:
import datetime

x = datetime.datetime.now()

print(x) # This prints present time

print(x.year)

print(x.strftime("%d")) # date of the month

print(x.strftime("%b")) # month 

print(x.strftime("%A")) # Day of week - full

print(x.strftime("%a")) # Day of week - short

y = datetime.datetime(2011, 4 , 2, 20, 30, 00) #Creating Date Objects

print(y)

2025-02-07 12:50:12.839980
2025
07
Feb
Friday
Fri
2011-04-02 20:30:00


#### **Built -in Math**

In [23]:
x = max( -15, 58, 12, 1, 0)
x1 = min( -15, 58, 12, 1, 0)


print(x)
print(x1)

z1 = pow(2,3)
z2 = abs(-12.3)


print(z1)
print(z2)

58
-15
8
12.3


#### **Using math module**

In [24]:
import math

x = math.sqrt(64)

y = math.ceil(12.45)
z = math.floor(12.45)

z1 = math.pi
z2 = math.trunc(56.78)
z3 = math.modf(56.78)
print(x,y,z,z1,z2,z3)

8.0 13 12 3.141592653589793 56 (0.7800000000000011, 56.0)


In [26]:
# Factorial
factorial_value = math.factorial(5)

# Logarithm
log_value = math.log(100, 10)

print("Factorial of 5:", factorial_value)
print("Logarithm of 100 with base 10:", log_value)

Factorial of 5: 120
Logarithm of 100 with base 10: 2.0


# Python Date and Math Modules

## 1. `datetime` Module

The `datetime` module supplies classes for manipulating dates and times.

### Key Concepts
- **Datetime Object**: Represents a specific point in time.
- **Time Object**: Represents a time of day (no date).
- **Date Object**: Represents a calendar date.
- **Timedelta Object**: Represents the difference between two dates or times.

### Common Classes
| Class Name         | Description                                                   |
|--------------------|---------------------------------------------------------------|
| `datetime.date`     | Represents the date (year, month, day).                       |
| `datetime.time`     | Represents the time (hour, minute, second, microsecond).      |
| `datetime.datetime` | Represents both date and time.                               |
| `datetime.timedelta`| Represents the difference between two `datetime` objects.    |

### Methods and Functions

- **`datetime.datetime.now()`**: Returns the current local date and time.
- **`datetime.datetime.today()`**: Returns the current local date and time.
- **`datetime.date.today()`**: Returns the current local date.
- **`datetime.datetime.strptime()`**: Parses a string into a datetime object.
- **`datetime.datetime.strftime()`**: Converts a datetime object to a string.
- **`datetime.timedelta(days=x)`**: Represents a difference of `x` days.





In [20]:
from datetime import datetime, timedelta

# Get current date and time
now = datetime.now()

# Format datetime
formatted_now = now.strftime("%Y-%m-%d %H:%M:%S")

# Add 5 days
five_days_later = now + timedelta(days=5)

print("Current Date and Time:", formatted_now)
print("Five Days Later:", five_days_later)

Current Date and Time: 2025-02-07 12:46:22
Five Days Later: 2025-02-12 12:46:22.523807


## 2. `math` Module

The `math` module provides mathematical functions.

### Key Functions

| Function            | Description                                                    |
|---------------------|----------------------------------------------------------------|
| `math.sqrt(x)`       | Returns the square root of `x`.                                |
| `math.pow(x, y)`     | Returns `x` raised to the power of `y`.                        |
| `math.sin(x)`        | Returns the sine of `x` (x in radians).                       |
| `math.cos(x)`        | Returns the cosine of `x` (x in radians).                     |
| `math.factorial(x)`  | Returns the factorial of `x`.                                  |
| `math.log(x, base)`  | Returns the logarithm of `x` to the given `base`.              |
| `math.radians(x)`    | Converts angle `x` from degrees to radians.                    |
| `math.degrees(x)`    | Converts angle `x` from radians to degrees.                    |
| `math.ceil(x)`       | Returns the smallest integer greater than or equal to `x`.     |
| `math.floor(x)`      | Returns the largest integer less than or equal to `x`.        |
| `math.trunc(x)`      | Returns the truncated integer value of `x`.                   |
| `math.modf(x)`       | Returns the fractional and integer parts of `x`.              |




## 3. Interview Questions

### Date Module Interview Questions

1. **What is the difference between `datetime.date` and `datetime.datetime`?**
   - `datetime.date` represents just the date (year, month, day), while `datetime.datetime` includes both date and time (year, month, day, hour, minute, second).

2. **How would you find the difference between two dates?**
   - Use the `datetime` objects and subtract them. The result will be a `timedelta` object, which can be used to get the difference in days or seconds.
   ```python
   from datetime import datetime
   date1 = datetime(2025, 2, 7)
   date2 = datetime(2025, 1, 1)
   diff = date1 - date2
   print(diff.days)  # Output will be the number of days


3. **How do you `format` a datetime object?**

    - Use the strftime() method. For example: datetime_object.strftime("%Y-%m-%d").

4. **How do you add or subtract days from a datetime object?**

    - Use datetime.timedelta() to add or subtract days from a datetime object.

In [1]:
from datetime import datetime, timedelta
today = datetime.today()
ten_days_later = today + timedelta(days=10)
print("Ten Days Later:", ten_days_later)


Ten Days Later: 2025-02-17 11:58:28.801715


5.  **How do you parse a `string` into a datetime object?**

    - Use the  `strptime() ` method to parse a string and convert it into a datetime object.

In [2]:
from datetime import datetime
date_string = "2025-02-07 14:30:00"
date_obj = datetime.strptime(date_string, "%Y-%m-%d %H:%M:%S")
print("Parsed Date:", date_obj)


Parsed Date: 2025-02-07 14:30:00


### Math Module Interview Questions

1. **What is the difference between `math.pow(x, y)` and `x ** y`?**
   - `math.pow(x, y)` returns a float, while `x ** y` returns the result as an integer or float depending on the operands.

2. **How do you calculate the square root of a number in Python?**
   - Use `math.sqrt(x)` or `x ** 0.5` to calculate the square root.

3. **What is the purpose of `math.factorial()`?**
   - It calculates the factorial of a given number, i.e., the product of all positive integers up to that number.

4. **How can you convert radians to degrees in Python?**
   - Use `math.degrees(radians)` to convert radians to degrees.

5. **How can you compute the sine and cosine of an angle in Python?**
   - Use `math.sin(x)` and `math.cos(x)`, where `x` is the angle in radians.

6. **How do you find the logarithm of a number in Python?**
   - Use `math.log(x, base)` to compute the logarithm of `x` to the specified base.
   ```python
   import math
   log_value = math.log(100, 10)
   print("Logarithm of 100 with base 10:", log_value)


## 4. Useful Tables for Quick Reference

### Date Formats (for `strftime`)

| Directive   | Meaning                           | Example              |
|-------------|-----------------------------------|----------------------|
| `%Y`        | Year (4 digits)                   | 2025                 |
| `%m`        | Month (2 digits)                  | 02                   |
| `%d`        | Day of the month (2 digits)       | 07                   |
| `%H`        | Hour (24-hour clock)              | 14                   |
| `%M`        | Minute (2 digits)                 | 35                   |
| `%S`        | Second (2 digits)                 | 59                   |
| `%f`        | Microsecond (6 digits)            | 123456               |

---

### Trigonometric Functions (for `math`)

| Function     | Description                       |
|--------------|-----------------------------------|
| `math.sin(x)`| Sine of `x` (x in radians)        |
| `math.cos(x)`| Cosine of `x` (x in radians)      |
| `math.tan(x)`| Tangent of `x` (x in radians)     |
| `math.asin(x)`| Inverse sine of `x`              |
| `math.acos(x)`| Inverse cosine of `x`            |
| `math.atan(x)`| Inverse tangent of `x`           |

---

### Mathematical Constants (for `math`)

| Constant    | Description                                   |
|-------------|-----------------------------------------------|
| `math.pi`   | The mathematical constant π (approx. 3.14159) |
| `math.e`    | The mathematical constant e (approx. 2.71828) |
| `math.inf`  | Positive infinity                             |
| `math.nan`  | Represents "Not a Number"                     |

---

### Date Methods (for `datetime`)

| Method                           | Description                                                       |
|----------------------------------|-------------------------------------------------------------------|
| `datetime.now()`                 | Returns the current local date and time.                          |
| `datetime.today()`               | Returns the current local date and time.                          |
| `datetime.date.today()`          | Returns the current local date.                                   |
| `datetime.datetime.strptime()`   | Parses a string into a `datetime` object.                         |
| `datetime.datetime.strftime()`   | Converts a `datetime` object to a string.                         |
| `datetime.timedelta(days=x)`     | Represents a difference of `x` days.                              |
